src/server/gc_worker/compaction_filter.rs

```diff
@@ -154,133 +216,279 @@ struct WriteCompactionFilter {
+            }
+        } else if let Some(deleting_filtered) = self.deleting_filtered.as_mut() {
+            deleting_filtered.1.push((commit_ts, sequence));
+        }
+
+        if !filtered {
+            return Ok(CompactionFilterDecision::Keep);
+        }
+        self.filtered += 1;
+        self.handle_filtered_write(write)?;
+        let decision = if self.remove_older {
+            // Use `Decision::RemoveAndSkipUntil` instead of `Decision::Remove` to avoid
+            // leaving tombstones, which can only be freed at the bottommost level.
+            debug_assert!(commit_ts > 0);
+            let prefix = Key::from_encoded_slice(key_prefix);
+            let skip_until = prefix.append_ts((commit_ts - 1).into()).into_encoded();
+            CompactionFilterDecision::RemoveAndSkipUntil(skip_until)
+        } else {
+            CompactionFilterDecision::Remove
+        };
+        Ok(decision)
+    }
+
+    fn meet_a_key_in_filter(&mut self) {
+        if self.is_bottommost_level && self.deleting_filtered.is_none() {
+            // The compaction hasn't met the next MVCC-delete mark.
+            self.near_seek_distance += 1;
+        }
+    }
+
+    // It's possible that elder versions than a MVCC-delete mark occurs in a higher level.
+    // So those versions needs to be handled by scan + delete.
+    fn handle_delete_mark(&mut self) -> Result<(), String> {
+        let (mark, mut filtered) = match self.deleting_filtered.take() {
+            Some((x, y)) => (x, y.into_iter().peekable()),
+            None => return Ok(()),
+        };
+        debug_assert!(self.is_bottommost_level);
+
+        let mut valid = self.write_iter_seek_to(&mark)?;
         let mut write_iter = self.write_iter.take().unwrap();
         if valid && write_iter.key() == mark {
             // The delete mark itself is handled in `filter`.
-            valid = write_iter.next().unwrap();
+            valid = write_iter.next()?;
         }
 
         while valid {
             let (key, value) = (write_iter.key(), write_iter.value());
-            if truncate_ts(key) != self.key_prefix.as_slice() {
+            if truncate_ts(key)? != self.key_prefix.as_slice() {
                 break;
             }
-            self.handle_filtered_write(parse_write(value));
-            self.delete_write_key(key);
-            valid = write_iter.next().unwrap();
+            let (_, current_ts) = split_ts(key)?;
+            let mut need_delete = true;
+            while let Some((ts, filtered_seqno)) = filtered.peek() {
+                match current_ts.cmp(ts) {
+                    CmpOrdering::Less => break,
+                    CmpOrdering::Equal => {
+                        // It's possible that there are multiple `key_ts` in different levels.
+                        // A potential case is that a snapshot has been ingested.
+                        let seq = write_iter.sequence().unwrap();
+                        assert!(seq >= *filtered_seqno);
```

Comment:
```suggestion
                        debug_assert!(seq >= *filtered_seqno);
```

Reply:
Instead of assert, shall we set `encountered_errors` instead?

Reply:
I prefer `debug_assert` because it's an impossible path.

Reply:
many things are impossible if nothing goes buggy. just want to be on the safe side.